docs/Using-SRIOV-and-Clear-Containers.md

```diff
@@ -0,0 +1,218 @@
+changes are needed, and how to make these changes.
+
+### IOMMU Groups and PCIe Access Control Services
+Taking a look at how the iommu groups are setup on your host system can help
+provide information on whether or not your NIC is setup appropriately with
+respect to PCIe Access Control Services. More specifically, if the PCI bridge is
+within the same iommu-group as your NIC, then this is an indication that either
+your device doesn't support ACS or that it doesn't share this capability
+appropriately by default.
+
+For example, when you run the following command, if all is setup properly, you
+should have the PCI for each ACS enabled NIC port in its own iommu_group. If
+you don't see any output when running the following, then you likely need to
+update your host's kernel config to disable VFNOIOMMU.
+
+```
+find /sys/kernel/iommu_groups/ -type l
+```
+
+For more details, see http://vfio.blogspot.com/2014/08/iommu-groups-inside-and-out.html
+
+### Updating the Host Kernel
+
+These directions are provided as an example, and are based on Ubuntu 16.04. The
+user will need to do equivalent for their distribution.
+
+1.  Grab kernel sources
+
+```
+sudo apt-get install linux-source-4.10.0
+sudo apt-get install linux-headers-4.10.0
+cd /usr/src/linux-source-4.10.0/
+sudo tar -xvf linux-source-4.10.0.tar.bz2
+cd linux-source-4.10.0
+sudo apt-get install libssl-dev
+```
+
+2. Check the config and update if necessary
+
+```
+sudo cp /boot/config-4.8.0-36-generic .config
+sudo make olddefconfig #and verify resulting .config does not have NOIOMMU set; ie: # CONFIG_VFIO_NOIOMMU is not set
+```
+
+3.  If necessary, add PCI Quirk for SRIOV NIC
+
+Now, depending on how your NIC describes its ACS capabilities, you may need to add
+a quirk in order to indicate that the given NIC does properly support ACS.  An
+example is given below, but your mileage will vary (at the very least, check your
+PCI-ID).
+
+```
+modify drivers/pci/quirks.c:
+line 4118:
+static const u16 pci_quirk_intel_pch_acs_ids[] = {
++        0x0c01,
+        /* Ibexpeak PCH */
+        0x3b42, 0x3b43, 0x3b44, 0x3b45, 0x3b46, 0x3b47, 0x3b48, 0x3b49,
+```
+
+4.  Build and install the kernel:
+
+```
+sudo make -j #and go get coffee
+sudo make modules -j 3
+sudo make modules_install
+sudo make install
+```
+
+5.  Edit grub to enable intel-iommu:
+
+```
+edit /etc/default/grub and add intel_iommu=on to cmdline:
+- GRUB_CMDLINE_LINUX=""
++ GRUB_CMDLINE_LINUX="intel_iommu=on"
+sudo update-grub
+```
+
+6.  Reboot and verify
+Host system should be ready now -- reboot and verify that expected cmdline and
+kernel version is booted (look at /proc/cmdline and /proc/version)
+
+```
+sudo reboot
+```
+
+## Setting up SRIO-V Device
+
+All of the prior sections are needed once to prepare the SRIOV host systems.
+The following will be needed per boot in order to facilitate setting up a
+physical device's virtual functions.
+
+For SRIOV, a physical device can create up to `sriov_totalvfs` virtual 
```

Comment:
`sriov_totalvfs` hasn't been introduced yet - it's explained below so this needs to be mentioned here.